firebase-segmentation/src/main/java/com/google/firebase/segmentation/FirebaseSegmentation.java

```diff
@@ -55,7 +79,161 @@ public static FirebaseSegmentation getInstance(@NonNull FirebaseApp app) {
     return app.get(FirebaseSegmentation.class);
   }
 
-  Task<Void> setCustomInstallationId(String customInstallationId) {
-    return Tasks.forResult(null);
+  Task<Void> setCustomInstallationId(@Nullable String customInstallationId) {
+    if (customInstallationId == null) {
+      return clearCustomInstallationId();
+    }
+    return updateCustomInstallationId(customInstallationId);
+  }
+
+  /**
+   * Update custom installation id of the {@link FirebaseApp} on Firebase segmentation backend and
+   * client side cache.
+   *
+   * <pre>
+   *     The workflow is:
+   *         check diff against cache or cache status is not SYNCED
+   *                                 |
+   *                  get Firebase instance id and token
+   *                      |                       |
+   *                      |      update cache with cache status PENDING_UPDATE
+   *                      |                       |
+   *                    send http request to backend
+   *                                 |
+   *              on success: set cache entry status to SYNCED
+   *                                 |
+   *                               return
+   * </pre>
+   */
+  private Task<Void> updateCustomInstallationId(String customInstallationId) {
+    CustomInstallationIdCacheEntryValue cacheEntryValue = localCache.readCacheEntryValue();
+    if (cacheEntryValue != null
+        && cacheEntryValue.getCustomInstallationId().equals(customInstallationId)
+        && cacheEntryValue.getCacheStatus() == CustomInstallationIdCache.CacheStatus.SYNCED) {
+      // If the given custom installation id matches up the cached
+      // value, there's no need to update.
+      return Tasks.forResult(null);
+    }
+
+    Task<InstanceIdResult> instanceIdResultTask = firebaseInstanceId.getInstanceId();
+    Task<Boolean> firstUpdateCacheResultTask =
+        instanceIdResultTask.onSuccessTask(
+            instanceIdResult ->
+                localCache.insertOrUpdateCacheEntry(
+                    CustomInstallationIdCacheEntryValue.create(
+                        customInstallationId,
+                        instanceIdResult.getId(),
+                        CustomInstallationIdCache.CacheStatus.PENDING_UPDATE)));
+
+    // Start requesting backend when first cache update is done.
+    Task<Code> backendRequestResultTask =
+        firstUpdateCacheResultTask.onSuccessTask(
+            firstUpdateCacheResult -> {
+              if (firstUpdateCacheResult) {
+                String iid = instanceIdResultTask.getResult().getId();
+                String iidToken = instanceIdResultTask.getResult().getToken();
+                return backendServiceClient.updateCustomInstallationId(
```

Comment:
What if the device is offline, do you want to hold on to the new value and sync when device gets connected or do you want to throw?

Reply:
I have thought this case.

If the device is offline, I would
1) the backendServiceClient return a error
2) the cache status will remain in PENDING_UPDATE or PENDING_CLEAR
3) have a separate listener that retry the PENDINGs when device is back online.

WDYT?

Reply:
Sounds good.